libraries/appbase/application.cpp

```diff
@@ -389,16 +316,162 @@ bfs::path application::data_dir()const
    return my->_data_dir;
 }
 
-void application::add_program_options( const options_description& cli, const options_description& cfg )
-{
-   my->_app_options.add( cli );
-   my->_app_options.add( cfg );
-   my->_cfg_options.add( cfg );
-}
 
 const variables_map& application::get_args() const
 {
    return my->_args;
 }
 
+void application_factory::add_program_options( const options_description& cli )
+{
+   global_options.add( cli );
+}
+
+application_factory& app_factory() { return application_factory::get_app_factory(); }
+
+
+void application_factory::set_program_options()
+{
+
+   options_description app_cfg_opts( "Application Config Options" );
+   options_description app_cli_opts( "Application Command Line Options" );
+   app_cfg_opts.add_options()
+         ("plugin", bpo::value< vector<string> >()->composing(), "Plugin(s) to enable, may be specified multiple times");
+
+   app_cfg_opts.add_options()
+         ("external-plugin", bpo::value< vector<string> >()->composing(), "External plugin(s) to enable, may be specified multiple times");
+
+   app_cli_opts.add_options()
+         ("startup-apps", bpo::value<vector<string>>()->composing(), "Applications to load at startup");
+
+   app_cli_opts.add_options()
+         ("external-plugins-dir", bpo::value<bfs::path>()->default_value( "external-plugins" ), "Directory containing external/runtime-loadable plugins binaries (absolute path or relative to the program option data-dir/)");
+
+   app_cli_opts.add_options()
+         ("help,h", "Print this help message and exit.")
+         ("version,v", "Print version information.")
+         ("data-dir,d", bpo::value<bfs::path>()->default_value( "sophia_app_data" ), "Directory containing configuration files, blockchain data and external plugins")
+         ("config,c", bpo::value<bfs::path>()->default_value( "config.ini" ), "Main configuration file path (absolute path or relative to the data-dir/configs/)");
+
+   app_options.add(app_cfg_opts);
+   global_options.add(app_cli_opts);
+}
+
+map<string, application& > application_factory::initialize( int argc, char** argv, vector< string > _autostart_plugins, bool start_apps )
```

Comment:
Isn't it better const & to vector of strings?

Reply:
It is...